csrc/dd_arith.c

```diff
@@ -488,3 +494,244 @@ ddouble tanhq(ddouble a)
+        return (ispositiveq(x)) ? Q_ZERO : Q_PI;
+    }
+
+    if (equalqq(x, y)) {
+        return (ispositiveq(y)) ? Q_PI_4: negq(Q_3PI_4);
+    }
+
+    if (equalqq(x, negq(y))) {
+        return (ispositiveq(y)) ? Q_3PI_4 : negq(Q_PI_4);
+    }
+
+    ddouble r = sqrtq(addqq(sqrq(x), sqrq(y)));
+    ddouble xx = divqq(x, r);
+    ddouble yy = divqq(y, r);
+
+    /* Compute double precision approximation to atan. */
+    // ddouble z = (ddouble){atan2(y.hi, x.hi), 0.};
+    ddouble z = adddq(atan2(y.hi, x.hi), Q_ZERO);
+    ddouble sin_z, cos_z;
+
+    if (fabs(xx.hi) > fabs(yy.hi)) {
+        /* Use Newton iteration 1.  z' = z + (y - sin(z)) / cos(z)  */
+        sincos(z, &sin_z, &cos_z);
+        z = addqq(z, divqq(subqq(yy, sin_z), cos_z));
+    } else {
+
+        /* Use Newton iteration 2.  z' = z - (x - cos(z)) / sin(z)  */
+        sincos(z, &sin_z, &cos_z);
+        z = subqq(z, divqq(subqq(xx, cos_z), sin_z));
+    }
+
+  return z;
+}
+
+ddouble atan2dq(const double a, const ddouble b) {
+    return atan2qq(adddq(a, Q_ZERO), b);
+}
+
+ddouble atan2qd(const ddouble a, const double b) {
+    return atan2qq(a, adddq(b, Q_ZERO));
+}
+
+ddouble atanq(const ddouble a) {
+    return atan2qq(a, Q_ONE);
+}
+
+ddouble acosq(const ddouble a) {
+    ddouble abs_a = absq(a);
+
+    if (greaterqq(abs_a, Q_ONE)) {
+        return nanq();
+    }
+
+    if (isoneq(abs_a)) {
+        return (ispositiveq(a)) ? Q_ZERO : Q_PI;
+    }
+
+    return atan2qq(sqrtq(subdq(1.0, sqrq(a))), a);
+}
+
+ddouble asinq(const ddouble a) {
+    ddouble abs_a = absq(a);
+
+    if (greaterqd(abs_a, 1.0)) {
+        return nanq();
+    }
+
+    if (isoneq(abs_a)) {
+        return (ispositiveq(a)) ? Q_PI_2 : negq(Q_PI_2);
+    }
+
+    return atan2qq(a, sqrtq(subdq(1.0, sqrq(a))));
+}
+
+ddouble asinhq(const ddouble a) {
+    return logq(addqq(a,sqrtq(addqd(sqrq(a),1.0))));
+}
+
+ddouble acoshq(const ddouble a) {
+    if (lessqd(a, 1.0)) {
+        return nanq();
+    }
+
+    return logq(addqq(a, sqrtq(subqd(sqrq(a), 1.0))));
+}
+
+ddouble atanhq(const ddouble a) {
+    if (equalqd(a, -1.0)) {
+        return negq(infq());
+    } else if (equalqd(a, 1.0)) {
+        return infq();
+    } else if (greaterqd(absq(a), 1.0)) {
+        return nanq();
+    }
+
+    return mul_pwr2(logq(divqq(adddq(1.0, a) , subdq(1.0, a))), 0.5);
+}
+
+ddouble powqq(const ddouble a, const ddouble b) {
+    if (equalqq(a, Q_ZERO) && equalqq(b, Q_ZERO)) {
+        return Q_ONE;
+    } else if (equalqq(a, Q_ZERO) && !equalqq(b, Q_ZERO)) {
+        return Q_ZERO;
+    } else {
+        return expq(mulqq(b, logq(a)));
+    }
+}
+
+ddouble powqd(const ddouble a, const double b) {
+    if (equalqq(a, Q_ZERO) && equaldq(b, Q_ZERO)) {
+        return Q_ONE;
+    } else if (equalqq(a, Q_ZERO) && !equaldq(b, Q_ZERO)) {
+        return Q_ZERO;
+    } else {
+        return expq(muldq(b, logq(a)));
+    }
+}
+
+ddouble powdq(const double a, const ddouble b) {
+    powqq(adddq(a, Q_ZERO), b);
```

Comment:
Careful, you forgot the return!
```suggestion
    return powqq((ddouble){a, 0.0}, b);
```

Reply:
Nice catch! Since there is not much code here, I've changed it to deal with the `double` type directly instead of going through `powqq`.